### PR TITLE
RESTEASY-2274 make master branch to use RESTEasy 4.0.0.Final

### PR DESCRIPTION
**Pull request by liweinan**

### CODE REVIEW

File: resteasy-spring-boot-starter/src/main/java/org/jboss/resteasy/springboot/ResteasyAutoConfiguration.java [hunk 1]
```diff
@@ -39,7 +40,7 @@
     @Bean
     @Qualifier("ResteasyProviderFactory")
     public static BeanFactoryPostProcessor springBeanProcessor() {
-        ResteasyProviderFactory resteasyProviderFactory = new ResteasyProviderFactory();
+        ResteasyProviderFactory resteasyProviderFactory = new ResteasyProviderFactoryImpl();
```
Comment (asoldano) [inline resteasy-spring-boot-starter/src/main/java/org/jboss/resteasy/springboot/ResteasyAutoConfiguration.java]: @liweinan you should actually do ``ResteasyProviderFactory.newInstance()`` here.

File: pom.xml [hunk 2]
```diff
@@ -21,7 +21,7 @@
         <repository>
             <id>jboss-releases-repository</id>
             <name>JBoss Releases Repository</name>
-            <url>https://repository.jboss.org/nexus/service/local/staging/deploy/maven2/</url>
+            <url>https://repository.jboss.org/nexus/content/repositories/releases/</url>
```
Comment (asoldano) [inline pom.xml]: What's the reason for this?

File: resteasy-spring-boot-starter/pom.xml [hunk 3]
```diff
@@ -37,18 +37,12 @@
...
         <developerConnection>scm:git:git@github.com:resteasy/resteasy-spring-boot.git</developerConnection>
         <url>git@github.com:resteasy/resteasy-spring-boot.git</url>
     </scm>
-    <distributionManagement>
-        <repository>
-            <id>jboss-releases-repository</id>
-            <name>JBoss Releases Repository</name>
-            <url>https://repository.jboss.org/nexus/service/local/staging/deploy/maven2/</url>
-        </repository>
-    </distributionManagement>
```
Comment (asoldano) [inline resteasy-spring-boot-starter/pom.xml]: Same as above, what's the reason?

**Comment by sharathsrinivasa** [comment]
@liweinan Thanks for picking this up :) I was about to send this contribution today. Similar changes are done [here](https://github.com/sharathsrinivasa/resteasy-spring-boot/commits/upgrade-2-resteasy-4) for your reference. 

@asoldano @liweinan How about bumping the major version to 4.x inline with RestEasy? Any comments here?

**Comment by asoldano** [comment]
> @asoldano @liweinan How about bumping the major version to 4.x inline with RestEasy? Any comments here?

+1

**Comment by anilgursel** [comment]
`resteasy-spring-boot-starter-test` seems to be skipped since there were coverage issues earlier.  Once the CI runs `resteasy-spring-boot-starter-test`, `asyncRequestTest ` in `AsyncJobIT` would fail.  The problem was discussed in [https://github.com/resteasy/Resteasy/pull/2062#issuecomment-499984712](this Resteasy PR comment).

File: resteasy-spring-boot-starter/src/main/java/org/jboss/resteasy/springboot/ResteasyAutoConfiguration.java (hunk 1 above)
Comment (liweinan) [reply resteasy-spring-boot-starter/src/main/java/org/jboss/resteasy/springboot/ResteasyAutoConfiguration.java]: okay!

File: pom.xml (hunk 2 above)
Comment (liweinan) [reply pom.xml]: The `https://repository.jboss.org/nexus/service/local/staging/deploy/maven2` is our internal staging repo and needs jboss account to access:

image

The `https://repository.jboss.org/nexus/content/repositories/releases` is public:

image

Should we use the public one?

Comment (liweinan) [reply pom.xml]: In addition, we can remove the repository settings to use the released resteasy releases. wdyt?

File: resteasy-spring-boot-starter/pom.xml (hunk 3 above)
Comment (liweinan) [reply resteasy-spring-boot-starter/pom.xml]: Should we scatter these repository settings around?

Can we remove all these internal repository settings from the projects and make resteasy-spring-boot should consume the released RESTEasy version?

**Comment by liweinan** [comment]
@anilgursel Yes the tests are failing because of the update:

image

The tests need to be fixed for upgrade.

**Comment by liweinan** [comment]
@sharathsrinivasa Thanks for update! I see you have made some fixes in test cases related with multiple application tests. So I will ignore the tests you are working on and leave that to you after this PR, is that okay to you?

@asoldano My plan is firstly to upgrade master to use resteasy to 4.x(and created 3.x branch from 3.1.0.Final.

And then I'll ignore some failed tests to make this PR to be merged firstly, and leave it to team to fix them in future Pull Requests(because some are also related with RESTEasy side). Is that okay?

Here is the 3.x branch creation I have done:

image

Here is the branch: https://github.com/resteasy/resteasy-spring-boot/tree/3.x

File: resteasy-spring-boot-starter/src/main/java/org/jboss/resteasy/springboot/ResteasyAutoConfiguration.java (hunk 1 above)
Comment (liweinan) [reply resteasy-spring-boot-starter/src/main/java/org/jboss/resteasy/springboot/ResteasyAutoConfiguration.java]: okay!

File: resteasy-spring-boot-starter/pom.xml (hunk 3 above)
Comment (asoldano) [reply resteasy-spring-boot-starter/pom.xml]: @liweinan the reason why you're finding this in multiple poms is that they explicitly have the jboss-parent as parent. I don't think there's anything to refactor here, really, but if you think there is please consider doing that as part of a dedicated PR / jira.

**Comment by sharathsrinivasa** [comment]
>@sharathsrinivasa Thanks for update! I see you have made some fixes in test cases related with multiple application tests. So I will ignore the tests you are working on and leave that to you after this PR, is that okay to you?
@liweinan I added those as a means to reproduce the [bug ](https://github.com/resteasy/Resteasy/pull/2062). Once that bug is fixed, I hope, those unit tests would be auto fixed.

**Comment by liweinan** [comment]
@asoldano thanks for the info. I'll revert the changes on dependency managements.

@sharathsrinivasa okay!

I'll go on working on this PR, and let you know my progress.

**Comment by liweinan** [comment]
There are 3 tests failed, I start from finding the failure root cause of `ConfigurationIT.scanningTest()`.

**Comment by liweinan** [comment]
The `ConfigurationIT.scanningTest()` failed because of multiple jaxrs applications exists in the path.

And if I disable one of the application classes, the test will pass:

image

image

@asoldano @ronsigal I guess this is caused by some changes in RESTEasy 4? Do you have any comments on this?

**Comment by ronsigal** [comment]
Hey @liweinan,

In https://github.com/resteasy/Resteasy/pull/2077, I made a couple of changes to fix RESTEASY-2266 "Resources accessed under second ApplicationPath returns 404", which was about a problem with multiple Applications. Of the seven changed files, five are test related, and the changes to HttpServletDispatcher and ServletContainerDispatcher are quite simple. Maybe you could incorporate them and then test.

Actually, I'd like to just merge https://github.com/resteasy/Resteasy/pull/2077, but there are some failures I don't understand. I think they are travis problems, though. Could you take a look at that?

Thanks,
-Ron

**Comment by asoldano** [comment]
@liweinan @ronsigal I've merged resteasy/Resteasy#2077 .

**Comment by liweinan** [comment]
@ronsigal @asoldano okay! I'll test against our great master branch :D

**Comment by liweinan** [comment]
@asoldano @ronsigal After I upgrade resteasy dependency to 4.1.0-SNAPSHOT(which includes https://github.com/resteasy/Resteasy/pull/2077/files as Ron noted), the `ConfigurationIT.scanningTest()` passed. I'll go on working on rest of the tests.

**Comment by liweinan** [comment]
The two failed async tests are failed because the `asyncEnabled` is not set:

image

The test set the above property by this way:

image

But it seems its no longer working under newest version of RESTEasy. @ronsigal @asoldano Do you have any clue on this?

**Comment by liweinan** [comment]
In addition, in this way it can set the property:

image

Using spring context parameters does not work any more:

image

**Comment by liweinan** [comment]
I guess this is because we now use MicroProfile config:

image

Let me think about how to integrate it with `resteasy-spring-boot`.

**Comment by liweinan** [comment]
Okay I have read the document:

https://docs.jboss.org/resteasy/docs/4.0.0.Final/userguide/html_single/index.html#microprofile_config

image

And used system property:

image

And the test passed:

image

Great! :D

**Comment by asoldano** [comment]
@liweinan isn't setting a system-wide property too much of an expectation here, though?

**Comment by liweinan** [comment]
@asoldano 

From the documentation there are other ways to inject the property:

* Using config file: `Config config = new PropertiesConfigSource("file:/// ...");`
* Using CDI `@Inject`

Above two ways are not integrated with Springframework. Another way is to inject spring application xml config files.

For testings, for me it seems using system property is okay because Microprofile config accepts this way. wdyt? I can also change it to other ways if necessary.

btw, all the changes are done. The test failed because now this branch relies on `4.1.0-SNAPSHOT` which haven't released. My local tests all passed:

image

Could you please help to check if anything more need to be changed in this PR? Thanks!

**Comment by anilgursel** [comment]
@liweinan Setting `context-param`  `resteasy.async.job.service.enabled` is an option mentioned in Resteasy documentation, so it should be supported.  In addition, this feature used to be working.  There may be some other underlying issue in Resteasy configuration changes that might have impacted this test.

**Comment by liweinan** [comment]
@anilgursel okay! Let me check its underlying work scheme so see how to enable this feature again.

**Comment by liweinan** [comment]
For the async test failures, I have found the root cause. Here is the detail: In our resteasy 4.x, the `ConfigurationBootstrap` is written like this:

image

And this method is backed by `microprofile-config`: 

image

And it can't get the parameter from `servletContext`. On contrary, in resteasy 3.x, the logic is like this:

image

As the screenshot shown above, it's using the `ConfigurationBootstrap.getParameter()` method, and it's actually override by `ListenerBootstrap.getParameter()` method:

image

As we can see above, it's using the `servletContext.getInitParameter()` method to fetch the parameters. 

Only in above way can we get the parameter injected from `servlet-context`:

image

Nevertheless, I can see from `resteasy4.x` that `ListenerBootstrap` still have the `getInitParameter()` mtehod:

image

But it's no longer properly used by `ConfigurationBootstrap`. As shown above, it's using microprofile classes to fetch the parameter, and here is the verification of the root cause:

image

@asoldano @ronsigal Do you think this is a bug in resteasy 4.x or some usage problem?

**Comment by asoldano** [comment]
@liweinan , I don't have time to debug this; however AFAIR we have a MP Config source that relies on the servlet context, org.jboss.resteasy.microprofile.config.ServletContextConfigSource. Is it running? why is it not finding what you say should come from the servlet context? Maybe the servlet context is not pushed in the ResteasyContext as that MP Config source expects?
Please try to go deeper and find the root cause and fix (in resteasy-spring-boot or in resteasy) for this. Thanks.

**Comment by liweinan** [comment]
@asoldano Thanks for the information! Let me check `ServletContextConfigSource` firstly. Currently the `servletContext` is injected into `ListenerBootstrap` correctly:

541EDBA6-101A-431E-944B-A79C275317F7

image

But it’s not properly used:

image

I guess the design of `ConfigurationBootstrap` doesn’t assume the `servlet` scope, and the solution *would* be to override the `getParameter()` method in `ListenerBootstrap`. I’ll verify this and provide proper patch to `resteasy` side. Thanks for help!

**Comment by asoldano** [comment]
> I guess the design of ConfigurationBootstrap doesn’t assume the servlet scope, and the solution would be to override the getParameter() method in ListenerBootstrap

Or maybe the servlet context needs to be pushed into ResteasyContext... please check that too.

**Comment by liweinan** [comment]
@asoldano okay!

**Comment by liweinan** [comment]
All tests passed:

image

**Comment by liweinan** [comment]
I found `openjdk9` provided by `travis-ci` is already replaced by `openjdk11`:

image

I'll remove `openjdk9` and add `openjdk11` for testing.

**Comment by liweinan** [comment]
@asoldano @ronsigal The last thing is the versioning of the master branch. I guess we should move the version to `4.0.x` for master branch to keep sync with RESTEasy major version, wdyt?

If you agree, I'll release `4.0.0.Final` from this PR, and then move the master branch to `4.0.1-SNAPSHOT`, do you agree?

In additon, we have `3.x` branch here for `resteasy 3.x` adoption:

https://github.com/resteasy/resteasy-spring-boot/tree/3.x

**Comment by liweinan** [comment]
@asoldano okay!

**Pull request merged**